File: Examples/fetch15/first-15.py
#!/usr/bin/env python3
"""Fetch the given urls, at most 8 at a time, and stop once 15 have landed.

    python3 Examples/fetch15/first-15.py https://example.com https://...

The Bjolang, Clojure and F# versions beside this one all build a feeder, a crew
of workers and a channel between them, because that is what their concurrency
primitives are. Python's are a semaphore and a completion iterator, so the crew
here is a number and the channel is a for loop.
"""

import asyncio
import sys

import aiohttp

WANTED = 15
AT_ONCE = 8
TIMEOUT = aiohttp.ClientTimeout(total=15)


async def fetch(
    session: aiohttp.ClientSession, url: str, gate: asyncio.Semaphore
) -> str:
    """Read the whole body, holding one of the gate's slots while it runs."""
    async with gate, session.get(url) as response:
        await response.text()
    return url


async def first_to_land(urls: list[str], wanted: int, at_once: int) -> list[str]:
    """The first `wanted` urls to answer, in the order they answered.

    Fewer if too many fail: the loop ends when the requests do.
    """
    gate = asyncio.Semaphore(at_once)

    async with aiohttp.ClientSession(timeout=TIMEOUT) as session:
        pending = [asyncio.create_task(fetch(session, url, gate)) for url in urls]
        landed = []
        try:
            async for task in asyncio.as_completed(pending):
                try:
                    landed.append(await task)
                except (aiohttp.ClientError, TimeoutError, UnicodeDecodeError):
                    continue
                if len(landed) == wanted:
                    break
        finally:
            # Leaving the loop early leaves requests in flight, and the session
            # closes on the way out of the `async with`. Cancel them first, then
            # wait: without the second half those sockets tear down under it.
            for task in pending:
                task.cancel()
            await asyncio.gather(*pending, return_exceptions=True)
        return landed


async def main(urls: list[str]) -> None:
    for url in await first_to_land(urls, WANTED, AT_ONCE):
        print(url)


if __name__ == "__main__":
    asyncio.run(main(sys.argv[1:]))

File: Examples/fetch15/first-15.clj
;; first-15.clj — fetch the urls given, 8 at a time, stop after 15 have landed.
;;
;; The Clojure counterpart of first-15.bjo, in the same shape: one feeder
;; thread, a fixed crew of 8 workers, one collector (the calling thread).
;;
;; What plays the part of `with-cancel` is thread interruption. Every blocking
;; call the crew makes — `SynchronousQueue.take`, `.put` and `HttpClient.send`
;; — throws `InterruptedException`, so interrupting the crew is one call that
;; every fiber in flight hears. The `finally` in `fetch-first` is the scope: it
;; runs whether the collector returned or threw.
;;
;; Run it with:
;;
;;   clojure Examples/fetch15/first-15.clj https://example.com https://...

(import '[java.net URI]
        '[java.net.http HttpClient HttpRequest HttpResponse HttpResponse$BodyHandlers]
        '[java.time Duration]
        '[java.util.concurrent SynchronousQueue])

(def ^HttpClient client
  (-> (HttpClient/newBuilder)
      (.connectTimeout (Duration/ofSeconds 10))
      (.build)))

;; ---------------------------------------------------------------------------
;; One request
;; ---------------------------------------------------------------------------

;; Fetches the text at url. [:ok body] or [:err throwable].
;;
;; `InterruptedException` is rethrown rather than reported as an error: it is
;; the cancellation arriving, and it belongs to the worker's loop.
(defn- fetch-text [url]
  (try
    (let [request (-> (HttpRequest/newBuilder (URI/create url))
                      (.GET)
                      (.build))
          ^HttpResponse response (.send client request (HttpResponse$BodyHandlers/ofString))]
      [:ok (.body response)])
    (catch InterruptedException e (throw e))
    (catch Exception e [:err e])))

;; ---------------------------------------------------------------------------
;; The crew
;; ---------------------------------------------------------------------------

;; A virtual thread running f. Virtual because the crew is sized by how many
;; requests should be in flight rather than by how many threads are affordable,
;; and because every blocking call below releases its carrier thread.
(defn- spawn ^Thread [f]
  (Thread/startVirtualThread f))

;; Take a url, fetch it, hand the answer on, repeat. If fetch-text is an error,
;; go on to the next address.
(defn- worker [^SynchronousQueue jobs ^SynchronousQueue answers]
  (fn []
    (try
      (loop []
        (let [url (.take jobs)]
          (.put answers {:url url :outcome (fetch-text url)})
          (recur)))
      (catch InterruptedException _ nil))))

(defn- hire [jobs answers n]
  (doall (repeatedly n #(spawn (worker jobs answers)))))

;; Push every url at the crew. The queue is a rendezvous, so this parks on send
;; number 9 until a worker frees up.
(defn- feeder [^SynchronousQueue jobs urls]
  (fn []
    (try
      (doseq [url urls]
        (.put jobs url))
      (catch InterruptedException _ nil))))

;; Cancel the crew, then wait for it to actually finish. Without the join
;; `fetch-first` returns while eight sockets are still tearing down.
(defn- reap [threads]
  (doseq [^Thread t threads] (.interrupt t))
  (doseq [^Thread t threads] (.join t)))

;; ---------------------------------------------------------------------------
;; The whole thing
;; ---------------------------------------------------------------------------

;; `seen` counts everything that came back, `kept` only the successes. `seen` is
;; what makes this return if every request failed.
(defn- collect [^SynchronousQueue answers wanted total]
  (loop [seen 0
         kept []]
    (if (or (= (count kept) wanted) (= seen total))
      kept
      (let [answer (.take answers)]
        (if (= :ok (first (:outcome answer)))
          (recur (inc seen) (conj kept answer))
          (recur (inc seen) kept))))))

(defn fetch-first [urls wanted crew-size]
  (let [jobs (SynchronousQueue.)
        answers (SynchronousQueue.)
        crew (hire jobs answers crew-size)
        feed (spawn (feeder jobs urls))]
    (try
      (collect answers wanted (count urls))
      (finally
        (reap (cons feed crew))))))

(defn -main [& urls]
  (doseq [answer (fetch-first (vec urls) 15 8)]
    (println (:url answer))))

;; Loaded as a script rather than through a main-opt, so that it runs on a bare
;; clojure.main with no classpath to set up.
(apply -main *command-line-args*)
